src/transformers/models/canine/modeling_canine.py

```diff
@@ -0,0 +1,1745 @@
+    set to :obj:`True`. To be used in a Seq2Seq model, the model needs to initialized with both :obj:`is_decoder`
+    argument and :obj:`add_cross_attention` set to :obj:`True`; an :obj:`encoder_hidden_states` is then expected as an
+    input to the forward pass.
+    """
+
+    def __init__(self, config, add_pooling_layer=True):
+        super().__init__(config)
+        self.config = config
+        shallow_config = copy.deepcopy(config)
+        shallow_config.num_hidden_layers = 1
+
+        self.char_embeddings = CanineEmbeddings(config)
+        # shallow/low-dim transformer encoder to get a initial character encoding
+        self.initial_char_encoder = CanineEncoder(
+            shallow_config,
+            local=True,
+            always_attend_to_first_position=False,
+            first_position_attends_to_all=False,
+            attend_from_chunk_width=config.local_transformer_stride,
+            attend_from_chunk_stride=config.local_transformer_stride,
+            attend_to_chunk_width=config.local_transformer_stride,
+            attend_to_chunk_stride=config.local_transformer_stride,
+        )
+        self.chars_to_molecules = CharactersToMolecules(config)
+        # deep transformer encoder
+        self.encoder = CanineEncoder(config)
+        self.projection = ConvProjection(config)
+        # shallow/low-dim transformer encoder to get a final character encoding
+        self.final_char_encoder = CanineEncoder(shallow_config)
+
+        self.pooler = CaninePooler(config) if add_pooling_layer else None
+
+        self.init_weights()
+
+    def _prune_heads(self, heads_to_prune):
+        """
+        Prunes heads of the model. heads_to_prune: dict of {layer_num: list of heads to prune in this layer} See base
+        class PreTrainedModel
+        """
+        for layer, heads in heads_to_prune.items():
+            self.encoder.layer[layer].attention.prune_heads(heads)
+
+    def _create_3d_attention_mask_from_input_mask(self, from_tensor, to_mask):
+        """
+        Create 3D attention mask from a 2D tensor mask.
+
+        Args:
+            from_tensor: 2D or 3D Tensor of shape [batch_size, from_seq_length, ...].
+            to_mask: int32 Tensor of shape [batch_size, to_seq_length].
+
+        Returns:
+            float Tensor of shape [batch_size, from_seq_length, to_seq_length].
+        """
+        batch_size, from_seq_length = from_tensor.shape[0], from_tensor.shape[1]
+
+        to_seq_length = to_mask.shape[1]
+
+        to_mask = torch.reshape(to_mask, (batch_size, 1, to_seq_length)).float()
+
+        # We don't assume that `from_tensor` is a mask (although it could be). We
+        # don't actually care if we attend *from* padding tokens (only *to* padding)
+        # tokens so we create a tensor of all ones.
+        #
+        # `broadcast_ones` = [batch_size, from_seq_length, 1]
+        broadcast_ones = torch.ones(size=(batch_size, from_seq_length, 1), dtype=torch.float32)
```

Comment:
Hm, after some debugging it seems that `broadcast_ones` is on cpu device, whereas `from_tensor` and `to_mask` are on the "expected" device (e.g. cuda when `input_ids` was set to cuda device earlier).

Reply:
Hmm yeah I'm not testing the model on GPU currently, as my `torch.cuda.is_available()` prints `False`. I will have to get through the pain of updating CUDA.

Reply:
Fixed it, will push soon